src/app/core/metadata/metadata.service.ts

```diff
@@ -0,0 +1,390 @@
+import 'rxjs/add/operator/first'
+import 'rxjs/add/operator/take'
+
+import { Inject, Injectable } from '@angular/core';
+import {
+  ActivatedRoute,
+  Event,
+  NavigationEnd,
+  Params,
+  Router
+} from '@angular/router';
+import { Meta, MetaDefinition } from '@angular/platform-browser';
+
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+import { Observable } from 'rxjs/Observable';
+
+import { Bitstream } from '../shared/bitstream.model';
+import { CacheableObject } from '../cache/object-cache.reducer';
+import { DSpaceObject } from '../shared/dspace-object.model';
+import { Item } from '../shared/item.model';
+import { Metadatum } from '../shared/metadatum.model';
+import { ObjectCacheService } from '../cache/object-cache.service';
+import { RemoteDataBuildService } from '../cache/builders/remote-data-build.service';
+
+import { GLOBAL_CONFIG, GlobalConfig } from '../../../config';
+
+@Injectable()
+export class MetadataService {
+
+  private initialized: boolean;
+
+  private tagStore: Map<string, MetaDefinition[]>;
+
+  private currentObject: BehaviorSubject<DSpaceObject>;
+
+  constructor(
+    private router: Router,
+    private objectCacheService: ObjectCacheService,
+    private remoteDataBuildService: RemoteDataBuildService,
+    private meta: Meta,
+    @Inject(GLOBAL_CONFIG) private envConfig: GlobalConfig
+  ) {
+    // TODO: determine what open graph meta tags are needed and whether
+    // the differ per route. potentially add image based on DSpaceObject
+    this.meta.addTags([
+      { property: 'og:title', content: 'DSpace Angular Universal' },
+      { property: 'og:description', content: 'The modern front-end for DSpace 7.' }
+    ]);
+    this.initialized = false;
+    this.tagStore = new Map<string, MetaDefinition[]>();
+  }
+
+  public listenForRouteChange(): void {
+    this.router.events
+      .filter((event) => event instanceof NavigationEnd)
+      .map(() => this.router.routerState.root)
+      .map((route: ActivatedRoute) => {
+        route = this.getCurrentRoute(route);
+        return { params: route.params, data: route.data };
+      }).subscribe((routeInfo: any) => {
+        this.processRouteChange(routeInfo);
+      });
+  }
+
+  private processRouteChange(routeInfo: any): void {
+    this.clearMetaTags();
+    if (routeInfo.params.value.id && routeInfo.data.value.type) {
+      this.objectCacheService.getByUUID(routeInfo.params.value.id, routeInfo.data.value.type)
```

Comment:
I'm not sure reconstructing the object again from route params in this service is the way to go.

This way we're duplicating the code to fetch the main object the route is about which is quite complicated and not guaranteed to work the same way for every future route. 

To do so you also had to add things like normalized objects to the route data. In my opinion we should try to avoid accessing normalized objects and the ObjectCacheService directly everywhere but in the data services.

It makes more sense to me to use the object that's already fetched by the page components (e.g. [in the CommunityPageComponent](https://github.com/DSpace/dspace-angular/blob/master/src/app/+community-page/community-page.component.ts#L34)), and to have those page components set the currentObject in the MetadataService, or if you don't want to couple the page components with the MetadataService, perhaps create an intermediate service that holds the object(s) associated with the current route so any other service that might need it can fetch it there.

Reply:
I wanted to avoid a switch statement conditioning against the route in order to fetch the DSpaceObject.

Reply:
https://angular.io/api/platform-browser/Title

Setting the title is easy.

Reply:
I am ok with not using ObjectCacheService and RemoteDataBuildService if there was a way to get a serialized DSpaceObject from the DataService. I think it is to static if we had to condition to determine which service to use to retrieve the DSpaceObject. 

CommunityDataService, CollectionDataService, ItemDataService, and any future services. 

Also, if the ObjectCacheService and RemoteDataBuildService are not meant to be used by other services, they probably should not be injectable.

Reply:
The reason there's no general DSpaceObject dataservice is because there's currently no way to fetch a resource from the rest api without specifying its type. 

My idea for the dataservices was that they represent the dso endpoints on the rest api. The components that use them shouldn't need to know whether the resource they need is cached or not, they always call the find method in the same way, as if they were directly querying the rest api. 

That's why, even though we could write a service that fetches any old DSO from the cache regardless of its type, I haven't done so. Because you can't do that on the rest api and I want the cache to be invisible for parts of the code that don't directly need to interact with it.

The reason the ObjectCacheService etc are injectable is because they _do_ need to be injected in to some backend services, but ideally not everywhere. Perhaps that can be enforced by better use of modules.

Reply:
Do you suggest injecting CommunityDataService, CollectionDataService, and ItemDataService into the MetadataService and conditioning on the route to get the DSpaceObject from its corresponding service?

Reply:
Or is there an alternative? 

The reason I feel safe about using the ObjectCacheService is that the intended route will have already fetched its DSpaceObject. The route than defined the type and it knows how to retrieve what the component already requested.

Reply:
I didn't investigate the code, but I assume if two requests where made right after each other, the second would just piggy back on the first and wait for its response.

Reply:
>Do you suggest injecting CommunityDataService, CollectionDataService, and ItemDataService into the MetadataService and conditioning on the route to get the DSpaceObject from its corresponding service?

No, I suggest keeping them all out of MetadataService, and rather injecting MetadataService in to the page components, and setting the current object on MetadataService from the page component. Or if you want to avoid injecting MetadataService directly in to the page components, creating an intermediary service that knows which objects correspond to the current route. Have the page components set the objects on that service, and have MetadataService fetch them from that service.

Reply:
So be it.